cbmc/proofs/IotMqtt_PublishSync/IotMqtt_PublishSync_harness.c

```diff
@@ -0,0 +1,135 @@
+ * a unsigned integer overflow failure. However, developers have reported
+ * that the use of this overflow is part of the function implementation.
+ * In order to mirror _IotMqtt_NextPacketIdentifier behaviour and avoid
+ * spurious alarms, we stub out this function to always
+ * return a nondetermistic odd value.
+ */
+uint16_t _IotMqtt_NextPacketIdentifier( void ) {
+  uint16_t id;
+
+  /* Packet identifiers will follow the sequence 1,3,5...65535,1,3,5... */
+  __CPROVER_assume(id % 2 == 1);
+
+  return id;
+}
+
+/**
+ * We abstract this function for performance reasons. In its original
+ * implementation, CBMC ended up creating byte extracts for all possible
+ * objects, due to the polymorphic nature of the linked list. We assume
+ * that the function is memory safe, and we free and havoc the list to
+ * demonstrate that no subsequent code makes use of the values in the list.
+ */
+void IotListDouble_RemoveAllMatches( const IotListDouble_t * const pList,
+				     bool ( *isMatch )( const IotLink_t * const pOperationLink, void * pCompare ),
+				     void * pMatch,
+				     void ( *freeElement )( void * pData ),
+				     size_t linkOffset )
+{
+  free_IotMqttSubscriptionList( pList );
+  allocate_IotMqttSubscriptionList( pList, SUBSCRIPTION_COUNT_MAX-1 );
+}
+
+
+/**
+ * We abstract these functions to map the expected behavior of
+ * IotSemaphore_TimedWait. If IotSemaphore_Post is called, the
+ * semaphore is positive. IotSemaphore_TimedWait should always
+ * return true if the semaphore is positive at some time during the wait.
+ */
+static bool flagSemaphore;
+
+void IotSemaphore_Post( IotSemaphore_t * pSemaphore )
+{
+  assert(pSemaphore != NULL);
+  flagSemaphore = true;
+}
+
+bool IotSemaphore_TimedWait( IotSemaphore_t * pSemaphore,
+                             uint32_t timeoutMs )
+{
+  assert(pSemaphore != NULL);
+  return flagSemaphore;
+}
+
+void harness()
+{
+  /* assume a valid MQTT connection */
+  IotMqttConnection_t mqttConnection = allocate_IotMqttConnection( NULL );
+  __CPROVER_assume( mqttConnection != NULL );
+  __CPROVER_assume( mqttConnection->pNetworkInterface != NULL );
+  __CPROVER_assume( IS_STUBBED_NETWORKIF_SEND( mqttConnection->pNetworkInterface ) );
+  __CPROVER_assume( IS_STUBBED_NETWORKIF_DESTROY( mqttConnection->pNetworkInterface ) );
+  ensure_IotMqttConnection_has_lists( mqttConnection );
+  __CPROVER_assume(valid_IotMqttConnection( mqttConnection ));
+ 
+  /* assume a valid publish info */
+  IotMqttPublishInfo_t *pPublishInfo = allocate_IotMqttPublishInfo( NULL );
+  __CPROVER_assume( valid_IotMqttPublishInfo( pPublishInfo ) );
+  
+  /* assume unconstrained inputs */
+  uint32_t flags;
+  uint32_t timeoutMs;
+
+  /* initialize semaphore flag */
+  flagSemaphore = false;
+
+  /* function under verification */
+  IotMqtt_PublishSync( mqttConnection, /* always assume a valid connection */
+                       nondet_bool() ? NULL : pPublishInfo,
```

Comment:
Move this to the __CPROVER_assume. 

Could use IMPLIES( pPublishInfo != NULL, valid_...(pPublishInfo))

Reply:
Done